crates/matrix-sdk-crypto/src/store/locks.rs

```diff
@@ -89,53 +137,95 @@ impl CryptoStoreLock {
             lock_key,
             lock_holder,
             backoff: Arc::new(Mutex::new(WaitingTime::Some(Self::INITIAL_BACKOFF_MS))),
-            num_holders: Arc::new(Mutex::new(0)),
+            num_holders: Arc::new(0.into()),
+            locking_attempt: Arc::new(Mutex::new(())),
         }
     }
 
     /// Try to lock once, returns whether the lock was obtained or not.
     #[instrument(skip(self), fields(?self.lock_key, ?self.lock_holder))]
-    pub async fn try_lock_once(&self) -> Result<bool, CryptoStoreError> {
-        // Hold the num_holders lock for the entire's function lifetime, to avoid
-        // internal races if called in a reentrant manner.
-        let mut holders = self.num_holders.lock().await;
+    pub async fn try_lock_once(&self) -> Result<Option<CryptoStoreLockGuard>, CryptoStoreError> {
+        // Hold onto the locking attempt mutext for the entire lifetime of this
+        // function, to avoid multiple reentrant calls.
+        let mut _attempt = self.locking_attempt.lock().await;
 
         // If another thread obtained the lock, make sure to only superficially increase
         // the number of holders, and carry on.
-        if *holders > 0 {
+        if self.num_holders.load(atomic::Ordering::SeqCst) > 0 {
+            // Note: between the above load and the fetch_add below, another thread may
+            // decrement `num_holders`. That's fine because that means the lock
+            // was taken by at least one thread, and after this call it will be
+            // taken by at least one thread.
             trace!("We already had the lock, incrementing holder count");
-            *holders += 1;
-            return Ok(true);
+            self.num_holders.fetch_add(1, atomic::Ordering::SeqCst);
+            let guard = CryptoStoreLockGuard { num_holders: self.num_holders.clone() };
+            return Ok(Some(guard));
         }
 
-        let inserted = self
+        let acquired = self
             .store
-            .insert_custom_value_if_missing(&self.lock_key, self.lock_holder.as_bytes().to_vec())
+            .try_take_leased_lock(Self::LEASE_DURATION_MS, &self.lock_key, &self.lock_holder)
             .await?;
 
-        if inserted {
-            trace!("Successfully acquired lock through db write");
-            *holders += 1;
-            return Ok(true);
-        }
+        if acquired {
+            // This is the first time we've acquired the lock. We're going to spawn the task
+            // that will renew the lease.
+
+            // Clone data to be owned by the task.
+            let this = self.clone();
+
+            matrix_sdk_common::executor::spawn(async move {
```

Comment:
Are we sure we don't want to keep a handle to this spawned task, so that we can abort it in case of emergency?

Reply:
Yes, it's also useful to make sure we don't spawn it multiple times :+1: Added some code to handle that.